airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_source.py

```diff
@@ -0,0 +1,185 @@
+        if isinstance(queue_item, Exception):
+            yield from concurrent_stream_processor.on_exception(queue_item)
+
+        elif isinstance(queue_item, PartitionGenerationCompletedSentinel):
+            yield from concurrent_stream_processor.on_partition_generation_completed(queue_item)
+
+        elif isinstance(queue_item, Partition):
+            concurrent_stream_processor.on_partition(queue_item)
+        elif isinstance(queue_item, PartitionCompleteSentinel):
+            yield from concurrent_stream_processor.on_partition_complete_sentinel(queue_item)
+        elif isinstance(queue_item, Record):
+            # record
+            yield from concurrent_stream_processor.on_record(queue_item)
+        else:
+            raise ValueError(f"Unknown queue item type: {type(queue_item)}")
+
+    def _get_streams_to_read_from(
+        self, catalog: ConfiguredAirbyteCatalog, logger: logging.Logger, stream_name_to_instance: Mapping[str, AbstractStream]
+    ) -> List[AbstractStream]:
+        """
+        Iterate over the configured streams and return a list of streams to read from.
+        If a stream is not configured, it will be skipped.
+        If a stream is configured but does not exist in the source and self.raise_exception_on_missing_stream is True, an exception will be raised
+        If a stream is not available, it will be skipped
+        """
+        stream_instances_to_read_from = []
+        for configured_stream in catalog.streams:
+            stream_instance = stream_name_to_instance.get(configured_stream.stream.name)
+            if not stream_instance:
+                if not self.raise_exception_on_missing_stream:
+                    continue
+                raise KeyError(
+                    f"The stream {configured_stream.stream.name} no longer exists in the configuration. "
+                    f"Refresh the schema in replication settings and remove this stream from future sync attempts."
+                )
+            else:
+                stream_availability = stream_instance.check_availability()
+                if not stream_availability.is_available():
+                    logger.warning(
+                        f"Skipped syncing stream '{stream_instance.name}' because it was unavailable. {stream_availability.message()}"
+                    )
+                    continue
+                stream_instances_to_read_from.append(stream_instance)
+        return stream_instances_to_read_from
+
+    def _streams_as_abstract_streams(self, config: Mapping[str, Any]) -> List[AbstractStream]:
+        """
+        Ensures the streams are StreamFacade and returns the underlying AbstractStream.
+        This is necessary because AbstractSource.streams() returns a List[Stream] and not a List[AbstractStream].
+        :param config:
+        :return:
+        """
+        streams = self.streams(config)
+        streams_as_abstract_streams = []
+        for stream in streams:
```

Comment:
This causes the solutions we are aiming for to not be applicable. One example is https://github.com/airbytehq/airbyte/issues/32308. 

One potential solution could be to have a CompositeSource that has multiple source each with exclusive catalogs. 
```
class CompositeSource(AbstractSource):
  def __init__(self, sources_and_catalogs: Tuple[AbstractSource, ConfiguredAirbyteCatalog]):
    ...
    
  def read(...):
    for source, catalog in self._sources_and_catalogs:
      yield from source.read(...)
```

If we think this is still relevant to support only some streams being concurrent but not all, we can explore this idea or others as well.